File: frontend/src/types/intelligenceRun.ts
export type BrandIntelligenceRunStatus =
  | 'not_started'
  | 'planning_questions'
  | 'waiting_scope_confirmation'
  | 'fetching_answers'
  | 'waiting_takeover'
  | 'analyzing_metrics'
  | 'building_world'
  | 'generating_recommendations'
  | 'waiting_user'
  | 'completed'
  | 'failed'
  | 'cancelled';

export interface BrandIntelligenceRun {
  id: string;
  entity_id: string;
  created_by_user_id?: string | null;
  origin_session_id?: string | null;
  analysis_task_id?: string | null;
  origin_surface: string;
  origin_event_id?: string | null;
  status: BrandIntelligenceRunStatus;
  stage: string;
  progress: number;
  message: string;
  run_goal: string;
  analysis_mode: string;
  dashboard_variant?: string | null;
  center_terms?: string[] | null;
  enabled_surfaces?: string[] | null;
  input_scope?: Record<string, unknown> | null;
  sample_scope?: Record<string, unknown> | null;
  output_refs?: Record<string, unknown> | null;
  requires_user_action: boolean;
  user_action_type?: string | null;
  blocking_reason?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  failed_at?: string | null;
  last_activity_at: string;
  created_at: string;
  updated_at: string;
}

export interface CreateBrandIntelligenceRunInput {
  run_goal?: string | null;
  analysis_mode?: string;
  input_scope?: Record<string, unknown> | null;
  origin_surface?: string;
  origin_session_id?: string | null;
  origin_event_id?: string | null;
  auto_dispatch?: boolean;
}

export interface ConfirmBrandIntelligenceRunInput {
  user_action_type?: string | null;
  feedback_text?: string | null;
  provided_inputs?: Record<string, unknown> | null;
  origin_event_id?: string | null;
}

export interface BrandIntelligenceRunResponse {
  run: BrandIntelligenceRun | null;
}

export const ACTIVE_BRAND_INTELLIGENCE_RUN_STATUSES: BrandIntelligenceRunStatus[] = [
  'not_started',
  'planning_questions',
  'waiting_scope_confirmation',
  'fetching_answers',
  'waiting_takeover',
  'analyzing_metrics',
  'building_world',
  'generating_recommendations',
  'waiting_user',
];

export function isActiveBrandIntelligenceRun(
  run?: BrandIntelligenceRun | null,
): boolean {
  return Boolean(run && ACTIVE_BRAND_INTELLIGENCE_RUN_STATUSES.includes(run.status));
}

/** M3: run is parked at the flow-plan confirmation gate (not yet executing). */
export function isAwaitingFlowPlanConfirmation(
  run?: BrandIntelligenceRun | null,
): boolean {
  if (!run) return false;
  if (run.status !== 'waiting_scope_confirmation') return false;
  const action = String(run.user_action_type || '');
  return (
    run.requires_user_action
    || action === 'flow_plan_confirmation'
    || action === 'confirm_flow_plan'
    || String(run.blocking_reason || '') === 'flow_plan_confirmation_required'
  );
}

/** True when the pipeline is actually executing (excludes plan confirmation). */
export function isExecutingBrandIntelligenceRun(
  run?: BrandIntelligenceRun | null,
): boolean {
  return isActiveBrandIntelligenceRun(run) && !isAwaitingFlowPlanConfirmation(run) && run?.status !== 'not_started';
}

File: aeo-platform/backend/app/schemas/intelligence_run.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.platform_fetch_methods import resolve_platform_fetch_methods


BrandIntelligenceRunStatusLiteral = Literal[
    "not_started",
    "planning_questions",
    "waiting_scope_confirmation",
    "fetching_answers",
    "waiting_takeover",
    "analyzing_metrics",
    "building_world",
    "generating_recommendations",
    "waiting_user",
    "completed",
    "failed",
    "cancelled",
]


class BrandIntelligenceRunCreate(BaseModel):
    run_goal: str | None = Field(default=None, max_length=1000)
    analysis_mode: str = Field(default="panorama", max_length=40)
    input_scope: dict[str, Any] | None = Field(default=None)
    origin_surface: str = Field(default="dashboard", max_length=80)
    origin_session_id: str | None = Field(default=None, max_length=80)
    origin_event_id: str | None = Field(default=None, max_length=160)
    auto_dispatch: bool = Field(default=True)


    @field_validator("input_scope")
    @classmethod
    def validate_platform_methods(cls, value):
        if isinstance(value, dict) and "platform_fetch_methods" in value:
            resolve_platform_fetch_methods(
                value["platform_fetch_methods"], platforms=value.get("platforms"),
                fetch_mode=value.get("fetch_mode", "fast"), explicit=True,
            )
        return value


class BrandIntelligenceRunConfirm(BaseModel):
    user_action_type: str | None = Field(default=None, max_length=80)
    feedback_text: str | None = Field(default=None, max_length=1000)
    provided_inputs: dict[str, Any] | None = Field(default=None)
    origin_event_id: str | None = Field(default=None, max_length=160)


class BrandIntelligenceRunResponse(BaseModel):
    id: str
    entity_id: str
    created_by_user_id: str | None = None
    origin_session_id: str | None = None
    analysis_task_id: str | None = None
    origin_surface: str
    origin_event_id: str | None = None
    status: BrandIntelligenceRunStatusLiteral
    stage: str
    progress: float
    message: str
    run_goal: str
    analysis_mode: str
    input_scope: dict[str, Any] | None = None
    sample_scope: dict[str, Any] | None = None
    output_refs: dict[str, Any] | None = None
    requires_user_action: bool
    user_action_type: str | None = None
    blocking_reason: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    failed_at: datetime | None = None
    last_activity_at: datetime
    created_at: datetime
    updated_at: datetime


class BrandIntelligenceRunEnvelope(BaseModel):
    run: BrandIntelligenceRunResponse | None
